subprojects/logging/src/main/java/org/gradle/internal/featurelifecycle/DeprecatedUsageBuildOperationProgressBroadcaster.java

```diff
@@ -17,31 +17,31 @@
 package org.gradle.internal.featurelifecycle;
 
 import org.gradle.internal.deprecation.DeprecatedFeatureUsage;
-import org.gradle.internal.operations.BuildOperationListener;
-import org.gradle.internal.operations.CurrentBuildOperationRef;
+import org.gradle.internal.operations.BuildOperationProgressEventEmitter;
 import org.gradle.internal.operations.OperationIdentifier;
 import org.gradle.internal.operations.OperationProgressEvent;
 import org.gradle.internal.time.Clock;
 
 public class DeprecatedUsageBuildOperationProgressBroadcaster {
 
     private final Clock clock;
-    private final BuildOperationListener listener;
+    private final BuildOperationProgressEventEmitter progressEventEmitter;
 
-    private final CurrentBuildOperationRef currentBuildOperationRef;
-
-    public DeprecatedUsageBuildOperationProgressBroadcaster(Clock clock, BuildOperationListener listener, CurrentBuildOperationRef currentBuildOperationRef) {
+    public DeprecatedUsageBuildOperationProgressBroadcaster(Clock clock, BuildOperationProgressEventEmitter progressEventEmitter) {
         this.clock = clock;
-        this.listener = listener;
-        this.currentBuildOperationRef = currentBuildOperationRef;
+        this.progressEventEmitter = progressEventEmitter;
     }
 
     void progress(DeprecatedFeatureUsage feature) {
-        OperationIdentifier id = currentBuildOperationRef.getId();
+        OperationIdentifier id = progressEventEmitter.getCurrentOperationIdentifier();
```

Comment:
It's a shame we discard events when the id is null. I'm guessing this is here because deprecated features may be used by threads that Gradle doesn't know about, which would include letting the user know that they shouldn't be using a thread that Gradle doesn't know about. Any chance we can have the scan plugin collect all deprecated usages, regardless of whether there is an associated build operation?

In the meantime, I wonder if a `emitNowIfCurrent(detail)` would be worthwhile, rather than having this consumer know how to construct all of the details for the event.

Reply:
> Any chance we can have the scan plugin collect all deprecated usages, regardless of whether there is an associated build operation?

The machinery currently requires a build operation for progress events. Short of changing that, we could just attach it to the root operation.